### PR TITLE
tokio-epoll-uring: use it on the layer-creating code paths

### PR DESCRIPTION
**Pull request by problame**
part of #6663 
See that epic for more context & related commits.

Problem
-------

Before this PR, the layer-file-creating code paths were using VirtualFile, but under the hood these were still blocking system calls.

Generally this meant we'd stall the executor thread, unless the caller "knew" and used the following pattern instead:

```
spawn_blocking(|| {
    Handle::block_on(async {
        VirtualFile::....().await;
    })
}).await
```

Solution
--------

This PR adopts `tokio-epoll-uring` on the layer-file-creating code paths in pageserver.

Note that on-demand downloads still use `tokio::fs`, these will be converted in a future PR.

Design: Avoiding Regressions With `std-fs` 
------------------------------------------

If we make the VirtualFile write path truly async using `tokio-epoll-uring`, should we then remove the `spawn_blocking`  + `Handle::block_on`  usage upstack in the same commit?

No, because if we’re still using the `std-fs` io engine, we’d then block the executor in those places where previously we were protecting us from that through the `spawn_blocking` .

So, if we want to see benefits from `tokio-epoll-uring` on the write path while also preserving the ability to switch between `tokio-epoll-uring`  and `std-fs` , where `std-fs`  will behave identical to what we have now, we need to ***conditionally* use `spawn_blocking + Handle::block_on`** .

I.e., in the places where we use that know, we’ll need to make that conditional based on the currently configured io engine.

It boils down to investigating all the places where we do `spawn_blocking(... block_on(... VirtualFile::...))`.

Detailed [write-up of that investigation in Notion](https://neondatabase.notion.site/Surveying-VirtualFile-write-path-usage-wrt-tokio-epoll-uring-integration-spawn_blocking-Handle-bl-5dc2270dbb764db7b2e60803f375e015?pvs=4
), made publicly accessible.

tl;dr: Preceding PRs addressed the relevant call sites:
-  `metadata` file: turns out we could simply remove it (#6777, #6769, #6775)
- `create_delta_layer()`: made sensitive to `virtual_file_io_engine` in #6986

NB: once we are switched over to `tokio-epoll-uring` everywhere in production, we can deprecate `std-fs`; to keep macOS support, we can use `tokio::fs` instead. That will remove this whole headache.

Code Changes In This PR
-----------------------

- VirtualFile API changes
  - `VirtualFile::write_at`
    - implement an `ioengine` operation and switch `VirtualFile::write_at` to it
  - `VirtualFile::metadata()`
    - curiously, we only use it from the layer writers' `finish()` methods
    - introduce a wrapper `Metadata` enum because `std::fs::Metadata` cannot be constructed by code outside rust std
  - `VirtualFile::sync_all()` and for completeness sake, add `VirtualFile::sync_data()`

Testing & Rollout
-----------------

Before merging this PR, we ran the CI with both io engines.

Additionally, the changes will soak in staging.

We could have a feature gate / add a new io engine `tokio-epoll-uring-write-path` to do a gradual rollout. However, that's not part of this PR.

Future Work
-----------

There's still some use of `std::fs` and/or `tokio::fs` for directory namespace operations, e.g. `std::fs::rename`. 

We're not addressing those in this PR, as we'll need to add the support in tokio-epoll-uring first. Note that rename itself is usually fast if the directory is in the kernel dentry cache, and only the fsync after rename is slow. These fsyncs are using tokio-epoll-uring, so, the impact should be small.

### CODE REVIEW

**Comment by problame** [comment]
While conducting self-review, I found that `create_delta_layer` needs to be bifurcated, transitioning back to draft mode until that's solved.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -474,7 +475,7 @@ jobs:
           TEST_RESULT_CONNSTR: ${{ secrets.REGRESS_TEST_RESULT_CONNSTR_NEW }}
           CHECK_ONDISK_DATA_COMPATIBILITY: nonempty
           BUILD_TAG: ${{ needs.tag.outputs.build-tag }}
-          PAGESERVER_VIRTUAL_FILE_IO_ENGINE: std-fs
+          PAGESERVER_VIRTUAL_FILE_IO_ENGINE: ${{ matrix.pageserver_virtual_file_io_engine }}
```
Comment (problame) [inline .github/workflows/build_and_test.yml:478]: Will revert this before merge. Depending on how far we are with the rollout to prod, I might change it from `std-fs` to `tokio-epoll-uring`.

**Review by koivunej — APPROVED** [review]
For me this is looking good.

**Review by VladLazar — APPROVED** [review]
